spec/unit/matrix-client.spec.ts

```diff
@@ -231,6 +287,130 @@ describe("MatrixClient", function() {
+            expect(path).toStrictEqual(
+                `/rooms/${encodeURIComponent(roomId)}/timestamp_to_event`,
+            );
+            expect(queryParams).toStrictEqual({
+                ts: '0',
+                dir: 'f',
+            });
+        });
+
+        it('should fallback to unstable endpoint when no support for stable endpoint', async () => {
+            httpLookups = [{
+                method: "GET",
+                path: `/rooms/${encodeURIComponent(roomId)}/timestamp_to_event`,
+                prefix: ClientPrefix.V1,
+                error: {
+                    httpStatus: 404,
+                    errcode: "M_UNRECOGNIZED",
+                },
+                expectQueryParams: {
+                    ts: '0',
+                    dir: 'f',
+                },
+            }, {
+                method: "GET",
+                path: `/rooms/${encodeURIComponent(roomId)}/timestamp_to_event`,
+                prefix: unstableMsc3030Prefix,
+                data: { event_id: eventId },
+                expectQueryParams: {
+                    ts: '0',
+                    dir: 'f',
+                },
+            }];
+
+            await client.timestampToEvent(roomId, 0, 'f');
+
+            expect(client.http.authedRequest.mock.calls.length).toStrictEqual(2);
+            const [
+                stableMethod,
+                stablePath,
+                stableQueryParams,
+                ,
+                { prefix: stablePrefix },
+            ] = client.http.authedRequest.mock.calls[0];
+            expect(stableMethod).toStrictEqual('GET');
+            expect(stablePrefix).toStrictEqual(ClientPrefix.V1);
+            expect(stablePath).toStrictEqual(
+                `/rooms/${encodeURIComponent(roomId)}/timestamp_to_event`,
+            );
+            expect(stableQueryParams).toStrictEqual({
+                ts: '0',
+                dir: 'f',
+            });
+
+            const [
+                unstableMethod,
+                unstablePath,
+                unstableQueryParams,
+                ,
+                { prefix: unstablePrefix },
+            ] = client.http.authedRequest.mock.calls[1];
+            expect(unstableMethod).toStrictEqual('GET');
+            expect(unstablePrefix).toStrictEqual(unstableMsc3030Prefix);
+            expect(unstablePath).toStrictEqual(
+                `/rooms/${encodeURIComponent(roomId)}/timestamp_to_event`,
+            );
+            expect(unstableQueryParams).toStrictEqual({
+                ts: '0',
+                dir: 'f',
+            });
+        });
+
+        it('should not fallback to unstable endpoint when stable endpoint returns an error', async () => {
+            httpLookups = [{
+                method: "GET",
+                path: `/rooms/${encodeURIComponent(roomId)}/timestamp_to_event`,
+                prefix: ClientPrefix.V1,
+                error: {
+                    httpStatus: 500,
```

Comment:
for safety, we should also test 429, 502, and 400

Reply:
Are you suggesting that we shouldn't fallback to the unstable endpoint for `429` (rate-limit) and `502` (bad-gateway, probably load balancer timeout type problem)? That seems reasonable 👍

For `400`, I don't think we can do the same thing yet given MSC3743 isn't a thing just yet. We still have to rely on that to show that there isn't support.

Reply:
> Are you suggesting that we shouldn't fallback to the unstable endpoint for `429` (rate-limit) and `502` (bad-gateway, probably load balancer timeout type problem)? That seems reasonable 👍

Diving into the tests/code again. We already do this. I don't see a benefit to adding extra tests for those particular status codes.

Reply:
Like above, we're testing that the specific handling we have in place is compliant with the MSC - a relatively important test imo.

Reply:
Fixed in 12cc7be31cd7f7891e1f89bdd1636d536e596e75

Reply:
While I appreciate that I didn't add the tests here, adding tests specifically for `429` and `502` feels like an extra burden we will have to maintain in the future. The shared `assertRequestsMade` test harness makes this burden of supporting any extra status code responses less though.

We already had tests for when we should fallback and a test to cover the generic case of *any other* status codes where we shouldn't fallback.

With `429`/`502`, we're covering the main scenarios we're likely to run into so it's not unreasonable but it's sliding into the territory of "why not have a test for every possible status code?"

Anyway, if no hearts and minds we're convinced ⏩

---

(adding the test for `400` alongside `404` for when we should fallback makes sense)